### PR TITLE
📝 vLLM-integration documentation

### PR DESCRIPTION
**Pull request by shirinyamani**
# What does this PR do?

Document on vllm-integration

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by qgallouedec** [comment]
The TOC doesn't render as expected, I think it's because you don't use the right heading levels 
Screenshot 2025-04-28 at 12 35 22

**Comment by qgallouedec** [comment]
It looks satisfactory on the whole. Can you proofread from start to finish to make sure there are no missing capitals, `` ` ``, links etc.? I've seen these elements missing in several places.
I'll leave it to @lewtun  or @edbeeching to conduct the rest of the reviews, as they'll probably have better advice than me on how to word things.

**Comment by shirinyamani** [comment]
> It looks satisfactory on the whole. Can you proofread from start to finish to make sure there are no missing capitals, `` ` ``, links etc.? I've seen these elements missing in several places.
> I'll leave it to @lewtun or @edbeeching to conduct the rest of the reviews, as they'll probably have better advice than me on how to word things.

@qgallouedec Okay I guess the rendering problem is resolved now + proofread✅

Screenshot 2025-04-28 at 2 35 40 PM

File: docs/source/vllm_integration.md [hunk 1]
```diff
@@ -1,14 +1,115 @@
 # vLLM Integration
+If you are here to learn about how to use vLLM with TRL, you are in the right place. This document will guide you through the process of using vLLM with TRL for faster generation in online methods like GRPO and Online DPO. We first summerize a tl;dr on how to use vLLM with TRL, and then we will go into the details of how it works under the hood. Let's go! 🔥
 
-<Tip warning={true}>
+## 🫠🚀 How can I use vLLM with TRL to make things go faster? TL;DR
+First run the server by; (This example allocate 4 GPUs for vLLM generation)
```
Comment (lewtun) [inline docs/source/vllm_integration.md:5]: turbo nit:
```suggestion
First run the server:
```

File: docs/source/vllm_integration.md [hunk 2]
```diff
@@ -1,14 +1,115 @@
...
+```sh
+trl vllm-serve --model Qwen/Qwen2.5-7B --tensor-parallel-size 2 --data-parallel-size 2
+```  
+Then, run the training script by passing `use_vllm=True` in the training arguments (this example allocate 4 GPUs for training) by;
```
Comment (lewtun) [inline docs/source/vllm_integration.md:9]:
```suggestion
In this example, we are sharding two copies of the model across 4 GPUs; increasing data parallelism increase throughput, while increasing tensor parallelism increases the size of models that can be served. Then, run the training script by passing `use_vllm=True` in the training arguments as follows:
```

File: docs/source/vllm_integration.md [hunk 3]
```diff
@@ -1,14 +1,115 @@
...
+  
+```sh
+CUDA_VISIBLE_DEVICES=4,5,6,7 accelerate launch train.py
+```  
+
+Sample of a simple `train.py` script:
```
Comment (lewtun) [inline docs/source/vllm_integration.md:15]: I suggest moving this up before the train command. Also, I would indicate whether you are using FSDP or ZeRO-2 here since I don't think a 7B model fits in DDP

File: docs/source/vllm_integration.md [hunk 4]
```diff
@@ -1,14 +1,115 @@
 # vLLM Integration
+If you are here to learn about how to use vLLM with TRL, you are in the right place. This document will guide you through the process of using vLLM with TRL for faster generation in online methods like GRPO and Online DPO. We first summerize a tl;dr on how to use vLLM with TRL, and then we will go into the details of how it works under the hood. Let's go! 🔥
```
Comment (edbeeching) [inline docs/source/vllm_integration.md:2]:
```suggestion
This document will guide you through the process of using vLLM with TRL for faster generation in online methods like GRPO and Online DPO. We first summerize a tl;dr on how to use vLLM with TRL, and then we will go into the details of how it works under the hood. Let's go! 🔥
```

File: docs/source/vllm_integration.md [hunk 5]
```diff
@@ -1,14 +1,115 @@
...
+    model="Qwen/Qwen2.5-7B",
+    args=training_args,
+    reward_funcs=reward_num_unique_chars,
+    train_dataset=dataset,
+)
+trainer.train()
+```
+
+## 🎬 Flashback: why do we need to use vLLM in online methods?
+Online methods like GRPO or Online DPO require the model to generate completions during training, which are then used to compute reward signals. However, generation can be extremely time-consuming, especially with large models. In the default setup (without vLLM), completions are generated using the [(unwrapped)model's `generate` method here](https://github.com/huggingface/trl/blob/f3e8c2304428ef16e9ae5de9e5741ed84d533b7b/trl/trainer/grpo_trainer.py#L965C39-L965C66). This approach quickly becomes a major bottleneck — generation is slow and inefficient, particularly for large batches or models. As a result, training times increase significantly, and overall efficiency drops. To address this, we turn to vLLM, which enables much faster and more scalable generation, helping eliminate this bottleneck in online methods.
```
Comment (edbeeching) [inline docs/source/vllm_integration.md:47]:
```suggestion
Online methods like GRPO or Online DPO require the model to generate completions during training, which are then used to compute reward signals. However, generation can be extremely time-consuming, especially with large models and reasoning models. In the default setup (without vLLM), completions are generated using the [(unwrapped)model's `generate` method here](https://github.com/huggingface/trl/blob/f3e8c2304428ef16e9ae5de9e5741ed84d533b7b/trl/trainer/grpo_trainer.py#L965C39-L965C66). This approach quickly becomes a major bottleneck — generation is slow and inefficient, particularly for large batches or models. As a result, training times increase significantly, and overall efficiency drops. To address this, we turn to vLLM, which enables much faster and more scalable generation, helping eliminate this bottleneck in online methods.
```

Comment (lewtun) [inline docs/source/vllm_integration.md:47]:
```suggestion
Online methods like GRPO or Online DPO require the model to generate completions during training, which are then used to compute reward signals. However, generation can be extremely time-consuming, especially with large models. In the default setup (without vLLM), completions are generated using the model's native `generate()` [method](https://github.com/huggingface/trl/blob/f3e8c2304428ef16e9ae5de9e5741ed84d533b7b/trl/trainer/grpo_trainer.py#L965C39-L965C66). However, this approach quickly becomes a major bottleneck — generation is slow and inefficient, particularly for large batches or models. As a result, training times increase significantly, and overall efficiency drops. To address this, we turn to vLLM, which enables much faster and more scalable generation, helping eliminate this bottleneck in online methods.
```

File: docs/source/vllm_integration.md [hunk 6]
```diff
@@ -1,14 +1,115 @@
...
+
+## 🤔 How does vLLM solve the slow generation issue?
+If you've ever done autoregressive decoder training, you know  all the input tokens to the LLM produce their attention key and value tensors, and these tensors are kept in GPU memory to later generate next tokens (Q) based on them. These cached key and value tensors are often referred to as KV cache.  However, this storing is really a pain as it occupies a lot of memory. So here is the secret sauce of vLLM, it uses a technique called PagedAttention to solve this problem. PagedAttention , which is inspired by the OS’s virtual memory concept stores continuous keys and values in **non-contiguous memory space** which is way more efficient. The detail of this is beyond the scope of this document, but in short, it allows the model to store the keys and values in a more efficient way, reducing the memory footprint and speeding up the generation process. If you are interested, make sure to check out the [vLLM PagedAttention](https://blog.vllm.ai/2023/06/20/vllm.html) for more details.
```
Comment (lewtun) [inline docs/source/vllm_integration.md:50]: Maybe we can also briefly mention that `vllm` also provides tensor parallelism, continuous batching, and optimised kernels which also help speed up generation
```suggestion
If you've ever done autoregressive decoder training, you know  all the input tokens to the LLM produce their attention key and value tensors, and these tensors are kept in GPU memory to later generate subsequent tokens based on them. These cached key and value tensors are often referred to as KV cache.  However, storing the KV cache occupies a lot of memory, so vLLM uses a technique called PagedAttention to solve this problem. PagedAttention , which is inspired by the OS’s virtual memory concept stores continuous keys and values in **non-contiguous memory space** which is way more efficient. The detail of this is beyond the scope of this document, but in short, it allows the model to store the keys and values in a more efficient way, reducing the memory footprint and speeding up the generation process. If you are interested, make sure to check out the [vLLM PagedAttention](https://blog.vllm.ai/2023/06/20/vllm.html) for more details.
```

File: docs/source/vllm_integration.md [hunk 7]
```diff
@@ -1,14 +1,115 @@
...
+
+
+## ⚙️ How to use vLLM in practice for generation in online methods in TRL?
```
Comment (lewtun) [inline docs/source/vllm_integration.md:53]: I wonder if this would be best merged with the quickstart above since it's possibly annoying for users to later discover they need to install the package

File: docs/source/vllm_integration.md [hunk 8]
```diff
@@ -1,14 +1,115 @@
...
+
+1. To use [vLLM](https://github.com/vllm-project/vllm), first install it using:
+
+```bash
+pip install vllm
```
Comment (lewtun) [inline docs/source/vllm_integration.md:58]: I think we should remove this recommendation since we need to pin `vllm` versions (i.e. just recommend `trl[vllm]`

File: docs/source/vllm_integration.md [hunk 9]
```diff
@@ -1,14 +1,115 @@
...
+
+2. Then, **start a vLLM server** by running:
+
+```bash
+trl vllm-serve --model <model_name>
+```
+## 🔎 What exactly happens when you run `trl vllm-serve --model <model_name>`?
+When you run for example `trl vllm-serve --model Qwen/Qwen2.5-7B --tensor-parallel-size 1 --data-parallel-size 4`, the following happens:
+![vllm](https://huggingface.co/datasets/trl-lib/documentation-images/resolve/main/vllm-doc.png)
+1. When you run a command like trl vllm-serve --model Qwen/Qwen2.5-7B --tensor-parallel-size 1 --data-parallel-size 4, vLLM first spawns multiple workers to handle incoming requests in parallel. The number of workers is determined by multiplying the --tensor-parallel-size and --data-parallel-size values. In this example, it spawns 4 workers (1 × 4).
```
Comment (lewtun) [inline docs/source/vllm_integration.md:81]:
```suggestion
1. vLLM first spawns multiple workers to handle incoming requests in parallel. The number of workers is determined by multiplying the `--tensor-parallel-size` and `--data-parallel-size` values. In this example, it spawns 4 workers (1 × 4).
```

Comment (edbeeching) [inline docs/source/vllm_integration.md:81]:
```suggestion
1. When you run a command like `trl vllm-serve --model Qwen/Qwen2.5-7B --tensor-parallel-size 1 --data-parallel-size 4`, vLLM first spawns multiple workers to handle incoming requests in parallel. The number of workers is determined by multiplying the `--tensor-parallel-size` and `--data-parallel-size` values. In this example, it spawns 4 workers (1 × 4).
```

File: docs/source/vllm_integration.md [hunk 10]
```diff
@@ -1,14 +1,115 @@
...
+
+3. Although the GPUs process requests independently and in parallel, they still need to communicate with each other. Remember that each GPU handles only a slice of the incoming prompts (for example, with 4 GPUs and 8 prompts using `--data-parallel-size=4`, each GPU processes 2 prompts).
+This GPU-to-GPU communication is managed efficiently by NVIDIA’s NCCL library. The communication mainly ensures that each GPU gets its correct portion of the incoming requests — it’s lightweight and doesn’t interfere with generation itself.
+Separately, the number of completions to generate per prompt is controlled by the num_generations setting in the GRPO config. For instance, if you set num_generations=2 (like the picture above), each prompt will have 2 completions. So, with 8 prompts and num_generations=2, you would end up with 16 completions total — regardless of the number of GPUs or parallelism settings.
+
+4. **🔬 How it works in practice in trl/grpo?**
+- The vLLM server starts by running the command: trl vllm-serve --model Qwen/Qwen2.5-7B.
```
Comment (edbeeching) [inline docs/source/vllm_integration.md:91]:
```suggestion
- The vLLM server starts by running the command: `trl vllm-serve --model Qwen/Qwen2.5-7B`.
```

File: docs/source/vllm_integration.md [hunk 11]
```diff
@@ -1,14 +1,115 @@
...
+- Once the server is running, it generates completions based on requests from the client (trainer) using vllm_client.generate here.
+- The client (trainer) then requests these completions from the server.
+- These completions are used to compute the reward signal.
+- Based on the reward signal and the model’s output, the loss is computed, and the backward pass is performed to update the model’s weights.
+- **Note**: The server only handles completion generation — it doesn’t train the model. Therefore, the model’s weights aren’t updated on the server. Once the backward pass is complete, the client sends the updated weights to the server using `vllm_client.update_named_param(name, param.data)`.
 
-## TRL vLLM server
+## 📝 Important vLLM Notes:
```
Comment (lewtun) [inline docs/source/vllm_integration.md:98]: turbo nit:
```suggestion
## 📝 Important vLLM notes
```

File: docs/source/vllm_integration.md [hunk 12]
```diff
@@ -1,14 +1,115 @@
...
+When using vLLM, ensure the gpus assigned for training and generation are separate to avoid resource conflicts. For instance, if you plan to use 4 GPUs for training and another 4 for vLLM generation, you can specify GPU allocation for training using `CUDA_VISIBLE_DEVICES`.  See the example below;
```
Comment (lewtun) [inline docs/source/vllm_integration.md:99]:
```suggestion
When using vLLM, ensure the GPUs assigned for training and generation are separate to avoid resource conflicts. For instance, if you plan to use 4 GPUs for training and another 4 for vLLM generation, you can specify GPU allocation for training using `CUDA_VISIBLE_DEVICES`.  See the example below;
```

**Review by edbeeching — APPROVED** [review]

File: docs/source/vllm_integration.md [hunk 13]
```diff
@@ -1,14 +1,115 @@
...
+
+3. Although the GPUs process requests independently and in parallel, they still need to communicate with each other. Remember that each GPU handles only a slice of the incoming prompts (for example, with 4 GPUs and 8 prompts using `--data-parallel-size=4`, each GPU processes 2 prompts).
+This GPU-to-GPU communication is managed efficiently by NVIDIA’s NCCL library. The communication mainly ensures that each GPU gets its correct portion of the incoming requests — it’s lightweight and doesn’t interfere with generation itself.
+Separately, the number of completions to generate per prompt is controlled by the num_generations setting in the GRPO config. For instance, if you set num_generations=2 (like the picture above), each prompt will have 2 completions. So, with 8 prompts and num_generations=2, you would end up with 16 completions total — regardless of the number of GPUs or parallelism settings.
+
+4. **🔬 How it works in practice in trl/grpo?**
```
Comment (lewtun) [inline docs/source/vllm_integration.md:90]: Move from itemized list to subsection?

**Review by lewtun — APPROVED** [review]
Thanks for adding this nice guide @shirinyamani ! Overall it looks good with some small nits from my part :)

**Review by qgallouedec — APPROVED** [review]
